Add bpf_load (from the kernel) and add userspace loading program and implement most of flowradar

mtbentley opened the pull request:
This requires kernel headers, so for now I'm assuming you have a linux checkout at ../linux.  

This has most of (the server side of) flowradar implemented.

TODO: add "total flow count" somewhere (probably len 1 percpu array?) so we can be sure to swap out maps every time it gets near 8000.

Commit:
Add bpf_load, from the kernel

This requires kernel headers, so for now I'm assuming you have a linux
checkout at ../linux.

TODO: check that there is a linux checkout at ../linux
TODO: is there a better way to get these headers?

Commit:
Add xpd-flowrader_user

This (should) load the xpd program to an interface
It also sets up a map pin for the 0-th map

TODO: actually pin all the maps, figure out how to read the data

Commit:
Makefile cleanup :o

Commit:
Compile out 8021{D,AQ} support, since it's not tested

Commit:
Add map pins for each count map

Commit:
Check an edge case

From https://github.com/netoptimizer/prototype-kernel/blob/72e473c723bc8dbc389b00d2ec631729444e4998/kernel/samples/bpf/xdp_ddos01_blacklist_user.c#L103

Commit:
Clean up compiler warnings

Commit:
Clean up Makefile a bit

Commit:
Add to readme

Commit:
Add tool to dump maps

Commit:
fixup

Commit:
Add cjson, and make dump_maps output json

Commit:
Implement most of flowradar

Commit:
fix warning

Commit:
reuse old maps if they exist

Commit:
Add host info and ability to clear maps

Currently, host info only contains the host number, used in the hash
function

Commit:
re-write hash

Commit:
Fix an error message

Commit:
Use per cpu maps

matthewbentley commented:
The output looks like
```
$ sudo ./dump_maps | jq .                                                  [±add-bpf_load ✓]
{
  "bloomfilter": {
    "0": {
      "0x106": "0x8000000000000",
      "0x166": "0x2000000000000",
      "0x1d6": "0x4000000000000000",
      "0x278": "0x400000000000000",
      "0x2b0": "0x800",
      "0x3c4": "0x8000000000"
    },
    "1": {
      "0xde": "0x40000000000000",
      "0x106": "0x8000000000000",
      "0x14d": "0x4000",
      "0x166": "0x2000000000000",
      "0x1d6": "0x4000000000000000",
      "0x1de": "0x1000",
      "0x278": "0x400000000000000",
      "0x27e": "0x800000000",
      "0x2b0": "0x800",
      "0x30c": "0x20000000000000",
      "0x320": "0x2000000000000000",
      "0x3c4": "0x8000000000"
    }
  },
  "flow_info": {
    "0": {
      "0xee7": "saddr=0x400000a,daddr=0x100000a,sport=0x1389,dport=0x963a,proto=0x6,flow_count=0x1,packet_count=0x2f00c",
      "0x53cb": "saddr=0x400000a,daddr=0x100000a,sport=0x1389,dport=0x963a,proto=0x6,flow_count=0x1,packet_count=0x2f00c",
      "0x61fa": "saddr=0x400000a,daddr=0x100000a,sport=0x1389,dport=0x963a,proto=0x6,flow_count=0x1,packet_count=0x2f00c",
      "0x8a7e": "saddr=0x400000a,daddr=0x100000a,sport=0x1389,dport=0x963a,proto=0x6,flow_count=0x1,packet_count=0x2f00c",
      "0xa671": "saddr=0x400000a,daddr=0x100000a,sport=0x1389,dport=0x963a,proto=0x6,flow_count=0x1,packet_count=0x2f00c",
      "0xbe73": "saddr=0x400000a,daddr=0x100000a,sport=0x1389,dport=0x963a,proto=0x6,flow_count=0x1,packet_count=0x2f00c"
    },
    "1": {
      "0xee7": "saddr=0x963a1389,daddr=0xc000106,sport=0x2f0,dport=0x0,proto=0x0,flow_count=0x0,packet_count=0xa00",
      "0x53cb": "saddr=0x963a1389,daddr=0xc000106,sport=0x2f0,dport=0x0,proto=0x0,flow_count=0x0,packet_count=0xa00",
      "0x61fa": "saddr=0x963a1389,daddr=0xc000106,sport=0x2f0,dport=0x0,proto=0x0,flow_count=0x0,packet_count=0xa00",
      "0x8a7e": "saddr=0x963a1389,daddr=0xc000106,sport=0x2f0,dport=0x0,proto=0x0,flow_count=0x0,packet_count=0xa00",
      "0xa671": "saddr=0x963a1389,daddr=0xc000106,sport=0x2f0,dport=0x0,proto=0x0,flow_count=0x0,packet_count=0xa00",
      "0xbe73": "saddr=0x963a1389,daddr=0xc000106,sport=0x2f0,dport=0x0,proto=0x0,flow_count=0x0,packet_count=0xa00"
    }
  },
  "eth_proto_count": {
    "0": {
      "0x0800": "0x40f1b"
    },
    "1": {
      "0x0800": "0x1b712",
      "0x0806": "0x3"
    }
  },
  "ip_proto_count": {
    "0": {
      "0x0006": "0x2f00c",
      "0x0001": "0x11f0f"
    },
    "1": {
      "0x0006": "0x14f81",
      "0x0001": "0x6791"
    }
  },
  "sport_count": {
    "0": {
      "0x1389": "0x2f00c"
    },
    "1": {
      "0x1389": "0x14f81"
    }
  },
  "dport_count": {
    "0": {
      "0x963a": "0x2f00c"
    },
    "1": {
      "0x9638": "0x2",
      "0x963a": "0x14f7f"
    }
  },
  "sip_count": {
    "0": {
      "0x400000a": "0x2f00c"
    },
    "1": {
      "0x400000a": "0x14f81"
    }
  },
  "dip_count": {
    "0": {
      "0x100000a": "0x2f00c"
    },
    "1": {
      "0x100000a": "0x14f81"
    }
  },
  "host_info": {
    "0": {
      "0x0": "host=0x2775"
    },
    "1": {
      "0x0": "host=0x2775"
    }
  }
}
```

Commit:
Fix makefile

dbgrigsby commented on Makefile:
```diff
@@ -19,9 +43,13 @@ clean: unload
 	sudo rm /var/run/netns/h1 || true
 	rm xdp-flowradar.o || true
 	rm test-hash || true
+	rm xdp-flowradar || true
```
Curious, why `rm X || true` insead of `rm -f X`?

dbgrigsby commented on Makefile:
```diff
@@ -1,12 +1,36 @@
+# NOTE: this requires a kernel with tools/lib/bpf bulit (for libbpf.so)
+LINUX_SOURCE=../linux
+# ^^ is there a better way to do this??
```
I don't see a way to do this that is simpler for a dev than the current without being slower (downloading relevant bits in the makefile) or more messy (requiring a specific subfolder of the kernel)

dbgrigsby commented on bpf_load.c:
```diff
@@ -0,0 +1,669 @@
+// SPDX-License-Identifier: GPL-2.0
```
Should add sourcing: https://github.com/MarvellEmbeddedProcessors/mainline-public/blob/master/samples/bpf/bpf_load.c

I spent a while reading this before realizing some of the comments were likely not by you

mtbentley replied:
It's actually from `linux/samples/bpf/bpf_load.[ch]`

dbgrigsby commented on xdp-flowradar_kern.c:
```diff
@@ -93,31 +128,26 @@ struct five_tuple {
-	hash = hash * a + ft->dport;
-	a = a * b;
-	hash = hash * a + ft->proto;
+#pragma unroll
```
Maybe this is a common C practice, but why are these loops specifically unrolled and not others?

mtbentley replied:
The `_kern.c` file gets compiled into the bpf program, which can't have loops (the verifier will reject any jump to a previous instruction).

Normally `#pragma unroll` would be rather unusual, but in this situation we need it.

Commit:
Address comments

Commit:
put the host first in the hash

The goal here is that a small change to the host will result in a less
predicatable change in the hash

Commit:
Remove todo comment for fixed todo :)

Commit:
Namespace pins by "host"

Commit:
Add pin prefix number

Commit:
rename analyze.py

Commit:
set rlimit (to allow more mem usage) and add a setup script

Commit:
import hash function into python

Usage:
1. `make py-c-hash`
2. python: `from cHash import c_hash`
3. python: `[hex(c_hash(0x1000a,0x100000a,0x1389,0xcb66,0x6,10101,i)) for i in range(6)]`

Commit:
fixup py-c-hash

Commit:
clean up make clean

Commit:
Readme

Commit:
don't clean up maps or namespace in make clean

Commit:
don't unload in clean

Commit:
Fix struct packing issue
